Using jmx in JAVA_OPTS make other java tools fail

I am trying to enable jmx on jenkins but when i do that i cannot use other java tools, like gradle, because they will inherit `JAVA_OPTS`.


**Dockerfile that reproduces the problem:**


```
FROM jenkins/jenkins:lts

USER root

RUN apt-get update && apt-get install -y gradle

ENV JAVA_OPTS "\
 -Dcom.sun.management.jmxremote.local.only=false\
 -Dcom.sun.management.jmxremote.port=8011\
 -Dcom.sun.management.jmxremote.rmi.port=8011\
 -Dcom.sun.management.jmxremote.authenticate=false\
 -Dcom.sun.management.jmxremote.ssl=false\
 -Dcom.sun.management.jmxremote.ssl.need.client.auth=false\
 -Dcom.sun.management.jmxremote.registry.ssl=false\
 -Dcom.sun.management.jmxremote.authenticate=false\
 -Dcom.sun.management.jmxremote=true\
 -Dcom.sun.management.jmxremote.host=0.0.0.0\
"
```

**Command:**

```
docker build . -t jenkins-with-gradle
docker exec -it $(docker run -d -p 8080:8080 -p 50000:50000 jenkins-with-gradle) gradle`
```

**Output:**

```
Error: JMX connector server communication error: service:jmx:rmi://0.0.0.0:8011
sun.management.AgentConfigurationError: java.rmi.server.ExportException: Port already in use: 8011; nested exception is:
	java.net.BindException: Address already in use (Bind failed)
	at sun.management.jmxremote.ConnectorBootstrap.exportMBeanServer(ConnectorBootstrap.java:800)
	at sun.management.jmxremote.ConnectorBootstrap.startRemoteConnectorServer(ConnectorBootstrap.java:468)
	at sun.management.Agent.startAgent(Agent.java:262)
	at sun.management.Agent.startAgent(Agent.java:452)
Caused by: java.rmi.server.ExportException: Port already in use: 8011; nested exception is:
	java.net.BindException: Address already in use (Bind failed)
	at sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:346)
	at sun.rmi.transport.tcp.TCPTransport.exportObject(TCPTransport.java:254)
	at sun.rmi.transport.tcp.TCPEndpoint.exportObject(TCPEndpoint.java:411)
	at sun.rmi.transport.LiveRef.exportObject(LiveRef.java:147)
	at sun.rmi.server.UnicastServerRef.exportObject(UnicastServerRef.java:236)
	at sun.management.jmxremote.ConnectorBootstrap$PermanentExporter.exportObject(ConnectorBootstrap.java:199)
	at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:146)
	at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:122)
	at javax.management.remote.rmi.RMIConnectorServer.start(RMIConnectorServer.java:404)
	at sun.management.jmxremote.ConnectorBootstrap.exportMBeanServer(ConnectorBootstrap.java:796)
	... 3 more
Caused by: java.net.BindException: Address already in use (Bind failed)
	at java.net.PlainSocketImpl.socketBind(Native Method)
	at java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:387)
	at java.net.ServerSocket.bind(ServerSocket.java:375)
	at java.net.ServerSocket.<init>(ServerSocket.java:237)
	at sun.management.jmxremote.ConnectorBootstrap$HostAwareSocketFactory.createServerSocket(ConnectorBootstrap.java:864)
	at sun.rmi.transport.tcp.TCPEndpoint.newServerSocket(TCPEndpoint.java:666)
	at sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:335)
	... 12 more
```

As a workaroud, i changed `jenkins.sh` to also read java options from `JENKINS_JMX_OPTS`:


```
  while IFS= read -r -d '' item; do
      java_opts_array+=( "$item" )
    fi
  done < <([[ $JENKINS_JMX_OPTS ]] && xargs printf '%s\0' <<<"$JENKINS_JMX_OPTS")
```


Or maybe just unsetting `JAVA_OPTS` before[ this line](https://github.com/jenkinsci/docker/blob/master/jenkins.sh#L35) would solve the problem.

--- a/README.md
+++ b/README.md
@@ -83,11 +83,13 @@ If you connect agents using web sockets (since Jenkins 2.217), the TCP agent por
 # Passing JVM parameters
 
 You might need to customize the JVM running Jenkins, typically to adjust [system properties](https://www.jenkins.io/doc/book/managing/system-properties/) or tweak heap memory settings.
-Use the `JAVA_OPTS` environment variable for this purpose :
+Use the `JAVA_OPTS` or `JENKINS_JAVA_OPTS` environment variables for this purpose :
 
 ```
-docker run --name myjenkins -p 8080:8080 -p 50000:50000 --env JAVA_OPTS=-Dhudson.footerURL=http://mycompany.com jenkins/jenkins:lts-jdk11
+docker run --name myjenkins -p 8080:8080 -p 50000:50000 --env JENKINS_JAVA_OPTS=-Dhudson.footerURL=http://mycompany.com jenkins/jenkins:lts-jdk11
 ```
+
+JVM options specifically for the Jenkins controller should be set through `JENKINS_JAVA_OPTS`, as other tools might also respond to the `JAVA_OPTS` environment variable.
 
 # Configuring logging
 
@@ -141,10 +143,10 @@ or as a parameter to docker,
 docker run --name myjenkins -p 8080:8080 -p 50001:50001 --env JENKINS_SLAVE_AGENT_PORT=50001 jenkins/jenkins:lts-jdk11
 ```
 
-**Note**: This environment variable will be used to set the port adding the
-[system property][https://www.jenkins.io/doc/book/managing/system-properties/] `jenkins.model.Jenkins.slaveAgentPort` to **JAVA_OPTS**.
+**Note**: This environment variable will be used to set the
+[system property](https://www.jenkins.io/doc/book/managing/system-properties/) `jenkins.model.Jenkins.slaveAgentPort`.
 
-> If this property is already set in **JAVA_OPTS**, then the value of
+> If this property is already set in **JAVA_OPTS** or **JENKINS_JAVA_OPTS**, then the value of
 `JENKINS_SLAVE_AGENT_PORT` will be ignored.
 
 # Installing more tools

--- a/jenkins.ps1
+++ b/jenkins.ps1
@@ -18,10 +18,10 @@ Get-ChildItem -Recurse -File -Path 'C:/ProgramData/Jenkins/Reference' | ForEach-
 if(($args.Count -eq 0) -or ($args[0] -match "^--.*")) {
 
   # read JAVA_OPTS and JENKINS_OPTS into arrays to avoid need for eval (and associated vulnerabilities)
-  $java_opts_array = $env:JAVA_OPTS -split ' '
+  $java_opts_array = ($env:JAVA_OPTS -split ' ') + ($env:JENKINS_JAVA_OPTS -split ' ')
 
   $agent_port_property='jenkins.model.Jenkins.slaveAgentPort'
-  if(![System.String]::IsNullOrWhiteSpace($env:JENKINS_AGENT_PORT) -and ($env:JAVA_OPTS -notmatch "$agent_port_property")) {
+  if(![System.String]::IsNullOrWhiteSpace($env:JENKINS_AGENT_PORT) -and !($java_opts_array -match "$agent_port_property")) {
     $java_opts_array += "-D`"$agent_port_property=$env:JENKINS_AGENT_PORT`""
   }
 

--- a/jenkins.sh
+++ b/jenkins.sh
@@ -11,14 +11,17 @@ find "${REF}" \( -type f -o -type l \) -exec bash -c '. /usr/local/bin/jenkins-s
 # if `docker run` first argument start with `--` the user is passing jenkins launcher arguments
 if [[ $# -lt 1 ]] || [[ "$1" == "--"* ]]; then
 
+  # shellcheck disable=SC2001
+  effective_java_opts=$(sed -e 's/^ $//' <<<"$JAVA_OPTS $JENKINS_JAVA_OPTS")
+
   # read JAVA_OPTS and JENKINS_OPTS into arrays to avoid need for eval (and associated vulnerabilities)
   java_opts_array=()
   while IFS= read -r -d '' item; do
     java_opts_array+=( "$item" )
-  done < <([[ $JAVA_OPTS ]] && xargs printf '%s\0' <<<"$JAVA_OPTS")
+  done < <([[ $effective_java_opts ]] && xargs printf '%s\0' <<<"$effective_java_opts")
 
   readonly agent_port_property='jenkins.model.Jenkins.slaveAgentPort'
-  if [ -n "${JENKINS_SLAVE_AGENT_PORT:-}" ] && [[ "${JAVA_OPTS:-}" != *"${agent_port_property}"* ]]; then
+  if [ -n "${JENKINS_SLAVE_AGENT_PORT:-}" ] && [[ "${effective_java_opts:-}" != *"${agent_port_property}"* ]]; then
     java_opts_array+=( "-D${agent_port_property}=${JENKINS_SLAVE_AGENT_PORT}" )
   fi
 

--- a/tests/runtime.Tests.ps1
+++ b/tests/runtime.Tests.ps1
@@ -63,45 +63,74 @@ Describe "[$TEST_TAG] test jenkins arguments" {
   }
 }
 
-Describe "[$TEST_TAG] create test container" {
-  It 'start container' {
-    $cmd = @"
-docker --% run -d -e JAVA_OPTS="-Duser.timezone=Europe/Madrid -Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\"" --name $SUT_CONTAINER -P $SUT_IMAGE
-"@
-    Invoke-Expression $cmd
-    $lastExitCode | Should -Be 0
-  }
-}
+Describe "[$TEST_TAG] passing JVM parameters" {
+  BeforeAll {
+    $tzSetting = '-Duser.timezone=Europe/Madrid'
+    $tzRegex = [regex]::Escape("Europe/Madrid")
 
-Describe "[$TEST_TAG] test container is running" {
-  It 'is running' {
-    # give time to eventually fail to initialize
-    Start-Sleep -Seconds 5
-    Retry-Command -RetryCount 3 -Delay 1 -ScriptBlock { docker inspect -f "{{.State.Running}}" $SUT_CONTAINER ; if($lastExitCode -ne 0) { throw('Docker inspect failed') } } -Verbose | Should -BeTrue
-  }
-}
+    $cspSetting = @'
+-Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\"
+'@
+    $cspRegex = [regex]::Escape("default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;")
 
-Describe "[$TEST_TAG] Jenkins is initialized" {
-  It 'is initialized' {
-    # it takes a while for jenkins to be up enough
-    Retry-Command -RetryCount 30 -Delay 5 -ScriptBlock { Test-Url $SUT_CONTAINER "/api/json" } -Verbose | Should -BeTrue
-  }
-}
+    function Start-With-Opts() {
+      Param (
+        [string] $JAVA_OPTS,
+        [string] $JENKINS_JAVA_OPTS
+      )
 
-Describe "[$TEST_TAG] JAVA_OPTS are set" {
-  It 'CSP value' {
-    $content = (Get-JenkinsWebpage $SUT_CONTAINER "/systemInfo").Replace("</tr>","</tr>`n").Replace("<wbr>", "").Split("`n") | Select-String -Pattern '<td class="pane">hudson.model.DirectoryBrowserSupport.CSP</td>' 
-    $content | Should -Match ([regex]::Escape("default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;"))
+      $cmd = "docker --% run -d --name $SUT_CONTAINER -P"
+      if ($JAVA_OPTS.length -gt 0) {
+        $cmd += " -e JAVA_OPTS=`"$JAVA_OPTS`""
+      }
+      if ($JENKINS_JAVA_OPTS.length -gt 0) {
+        $cmd += " -e JENKINS_JAVA_OPTS=`"$JENKINS_JAVA_OPTS`""
+      }
+      $cmd += " $SUT_IMAGE"
+
+      Invoke-Expression $cmd
+      $lastExitCode | Should -Be 0
+
+      # give time to eventually fail to initialize
+      Start-Sleep -Seconds 5
+      Retry-Command -RetryCount 3 -Delay 1 -ScriptBlock { docker inspect -f "{{.State.Running}}" $SUT_CONTAINER ; if($lastExitCode -ne 0) { throw('Docker inspect failed') } } -Verbose | Should -BeTrue
+
+      # it takes a while for jenkins to be up enough
+      Retry-Command -RetryCount 30 -Delay 5 -ScriptBlock { Test-Url $SUT_CONTAINER "/api/json" } -Verbose | Should -BeTrue
+    }
+
+    function Get-Csp-Value() {
+      return (Get-JenkinsWebpage $SUT_CONTAINER "/systemInfo").Replace("</tr>","</tr>`n").Replace("<wbr>", "").Split("`n") | Select-String -Pattern '<td class="pane">hudson.model.DirectoryBrowserSupport.CSP</td>' 
+    }
+
+    function Get-Timezone-Value() {
+      return (Get-JenkinsWebpage $SUT_CONTAINER "/systemInfo").Replace("</tr>","</tr>`n").Replace("<wbr>", "").Split("`n") | Select-String -Pattern '<td class="pane">user.timezone</td>' 
+    }
   }
 
-  It 'Timezone set' {
-    $content = (Get-JenkinsWebpage $SUT_CONTAINER "/systemInfo").Replace("</tr>","</tr>`n").Replace("<wbr>", "").Split("`n") | Select-String -Pattern '<td class="pane">user.timezone</td>' 
-    $content | Should -Match ([regex]::Escape("Europe/Madrid"))
-  }
-}
+  It 'passes JAVA_OPTS' {
+    Start-With-Opts -JAVA_OPTS "$tzSetting $cspSetting"
 
-Describe "[$TEST_TAG] cleanup container" {
-  It 'cleanup' {
+    Get-Csp-Value | Should -Match $cspRegex
+    Get-Timezone-Value | Should -Match $tzRegex
+  }
+
+  It 'passes JENKINS_JAVA_OPTS' {
+    Start-With-Opts -JENKINS_JAVA_OPTS "$tzSetting $cspSetting"
+
+    Get-Csp-Value | Should -Match $cspRegex
+    Get-Timezone-Value | Should -Match $tzRegex
+  }
+
+  It 'JENKINS_JAVA_OPTS overrides JAVA_OPTS' {
+    Start-With-Opts -JAVA_OPTS "$tzSetting -Dhudson.model.DirectoryBrowserSupport.CSP=\`"default-src 'self';\`"" -JENKINS_JAVA_OPTS "$cspSetting"
+
+    Get-Csp-Value | Should -Match $cspRegex
+    Get-Timezone-Value | Should -Match $tzRegex
+  }
+
+  AfterEach {
     Cleanup $SUT_CONTAINER | Out-Null
   }
 }
+

--- a/tests/runtime.bats
+++ b/tests/runtime.bats
@@ -8,6 +8,10 @@ IMAGE=${IMAGE:-debian_jdk17}
 SUT_IMAGE=$(get_sut_image)
 SUT_DESCRIPTION="${IMAGE}-runtime"
 
+teardown() {
+  cleanup "$(get_sut_container_name)"
+}
+
 @test "[${SUT_DESCRIPTION}] test version in docker metadata" {
   local version=$(get_jenkins_version)
   assert "${version}" docker inspect --format '{{ index .Config.Labels "org.opencontainers.image.version"}}' $SUT_IMAGE
@@ -69,14 +73,13 @@ SUT_DESCRIPTION="${IMAGE}-runtime"
   assert_equal "${output}" "UTF-8"
 }
 
-@test "[${SUT_DESCRIPTION}] create test container with Jenkins initialize and JAVA_OPTS are set" {
+# parameters are passed as docker run parameters
+start-jenkins-with-jvm-opts() {
   local container_name
   container_name="$(get_sut_container_name)"
   cleanup "${container_name}"
 
-  run docker run --detach --name "${container_name}" --publish-all \
-    --env JAVA_OPTS="-Duser.timezone=Europe/Madrid -Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\"" \
-    $SUT_IMAGE
+  run docker run --detach --name "${container_name}" --publish-all "$@" $SUT_IMAGE
   assert_success
 
   # Container is running
@@ -85,11 +88,40 @@ SUT_DESCRIPTION="${IMAGE}-runtime"
 
   # Jenkins is initialized
   retry 30 5 test_url /api/json
+}
 
-  # JAVA_OPTS are set
+get-csp-value() {
   local sed_expr='s/<wbr>//g;s/<td class="pane">.*<\/td><td class.*normal">//g;s/<t.>//g;s/<\/t.>//g'
-  assert 'default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;' \
-    bash -c "curl -fsSL --user \"admin:$(get_jenkins_password)\" $(get_jenkins_url)/systemInfo | sed 's/<\/tr>/<\/tr>\'$'\n/g' | grep '<td class=\"pane\">hudson.model.DirectoryBrowserSupport.CSP</td>' | sed -e '${sed_expr}'"
-  assert 'Europe/Madrid' \
-    bash -c "curl -fsSL --user \"admin:$(get_jenkins_password)\" $(get_jenkins_url)/systemInfo | sed 's/<\/tr>/<\/tr>\'$'\n/g' | grep '<td class=\"pane\">user.timezone</td>' | sed -e '${sed_expr}'"
+  bash -c "curl -fsSL --user \"admin:$(get_jenkins_password)\" $(get_jenkins_url)/systemInfo | sed 's/<\/tr>/<\/tr>\'$'\n/g' | grep '<td class=\"pane\">hudson.model.DirectoryBrowserSupport.CSP</td>' | sed -e '${sed_expr}'"
+}
+
+get-timezone-value() {
+  local sed_expr='s/<wbr>//g;s/<td class="pane">.*<\/td><td class.*normal">//g;s/<t.>//g;s/<\/t.>//g'
+  bash -c "curl -fsSL --user \"admin:$(get_jenkins_password)\" $(get_jenkins_url)/systemInfo | sed 's/<\/tr>/<\/tr>\'$'\n/g' | grep '<td class=\"pane\">user.timezone</td>' | sed -e '${sed_expr}'"
+}
+
+@test "[${SUT_DESCRIPTION}] passes JAVA_OPTS as JVM options" {
+  start-jenkins-with-jvm-opts --env JAVA_OPTS="-Duser.timezone=Europe/Madrid -Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\""
+
+  # JAVA_OPTS are used
+  assert 'default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;' get-csp-value
+  assert 'Europe/Madrid' get-timezone-value
+}
+
+@test "[${SUT_DESCRIPTION}] passes JENKINS_JAVA_OPTS as JVM options" {
+  start-jenkins-with-jvm-opts --env JENKINS_JAVA_OPTS="-Duser.timezone=Europe/Madrid -Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\""
+
+  # JENKINS_JAVA_OPTS are used
+  assert 'default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;' get-csp-value
+  assert 'Europe/Madrid' get-timezone-value
+}
+
+@test "[${SUT_DESCRIPTION}] JENKINS_JAVA_OPTS overrides JAVA_OPTS" {
+  start-jenkins-with-jvm-opts \
+    --env JAVA_OPTS="-Duser.timezone=Europe/Madrid -Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'\"" \
+    --env JENKINS_JAVA_OPTS="-Dhudson.model.DirectoryBrowserSupport.CSP=\"default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self' 'unsafe-inline';\""
+
+  # JAVA_OPTS and JENKINS_JAVA_OPTS are used
+  assert 'default-src &#039;self&#039;; script-src &#039;self&#039; &#039;unsafe-inline&#039; &#039;unsafe-eval&#039;; style-src &#039;self&#039; &#039;unsafe-inline&#039;;' get-csp-value
+  assert 'Europe/Madrid' get-timezone-value
 }

